### ISSUE TITLE
[MSOURCES-94] Heap space OOM

### ISSUE THREAD

**Issue by jira-importer** [apache/maven-source-plugin#166]
**[Stephan Schroevers](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=stephan202)** opened **[MSOURCES-94](https://issues.apache.org/jira/browse/MSOURCES-94?redirect=false)** and commented

After upgrading from version 2.4 to 3.0.0 our aggregate build (comprising 52 jar/war modules) started failing when run on Travis CI:

```
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-source-plugin:3.0.0:jar-no-fork (generate-source-jar) on project etl-executor-impl: Error creating source archive: Problem creating jar: Execution exception: Java heap space -> [Help 1]
org.apache.maven.lifecycle.LifecycleExecutionException: Failed to execute goal org.apache.maven.plugins:maven-source-plugin:3.0.0:jar-no-fork (generate-source-jar) on project etl-executor-impl: Error creating source archive: Problem creating jar: Execution exception
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:216)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:153)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:145)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject(LifecycleModuleBuilder.java:116)
    at org.apache.maven.lifecycle.internal.builder.multithreaded.MultiThreadedBuilder$1.call(MultiThreadedBuilder.java:188)
    at org.apache.maven.lifecycle.internal.builder.multithreaded.MultiThreadedBuilder$1.call(MultiThreadedBuilder.java:184)
    at java.util.concurrent.FutureTask.run(FutureTask.java:266)
    at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
    at java.util.concurrent.FutureTask.run(FutureTask.java:266)
    at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
    at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
    at java.lang.Thread.run(Thread.java:745)
Caused by: org.apache.maven.plugin.MojoExecutionException: Error creating source archive: Problem creating jar: Execution exception
    at org.apache.maven.plugins.source.AbstractSourceJarMojo.packageSources(AbstractSourceJarMojo.java:311)
    at org.apache.maven.plugins.source.AbstractSourceJarMojo.packageSources(AbstractSourceJarMojo.java:253)
    at org.apache.maven.plugins.source.AbstractSourceJarMojo.execute(AbstractSourceJarMojo.java:216)
    at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo(DefaultBuildPluginManager.java:132)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:208)
    ... 11 more
Caused by: org.codehaus.plexus.archiver.ArchiverException: Problem creating jar: Execution exception
    at org.codehaus.plexus.archiver.AbstractArchiver.createArchive(AbstractArchiver.java:982)
    at org.apache.maven.archiver.MavenArchiver.createArchive(MavenArchiver.java:650)
    at org.apache.maven.plugins.source.AbstractSourceJarMojo.packageSources(AbstractSourceJarMojo.java:303)
    ... 15 more
Caused by: java.io.IOException: Execution exception
    at org.codehaus.plexus.archiver.zip.AbstractZipArchiver.close(AbstractZipArchiver.java:807)
    at org.codehaus.plexus.archiver.AbstractArchiver.createArchive(AbstractArchiver.java:969)
    ... 17 more
Caused by: java.lang.OutOfMemoryError: Java heap space
    at org.codehaus.plexus.archiver.zip.ByteArrayOutputStream.needNewBuffer(ByteArrayOutputStream.java:121)
    at org.codehaus.plexus.archiver.zip.ByteArrayOutputStream.<init>(ByteArrayOutputStream.java:90)
    at org.codehaus.plexus.archiver.zip.OffloadingOutputStream.<init>(OffloadingOutputStream.java:109)
    at org.codehaus.plexus.archiver.zip.OffloadingOutputStream.<init>(OffloadingOutputStream.java:89)
    at org.codehaus.plexus.archiver.zip.DeferredScatterOutputStream.<init>(DeferredScatterOutputStream.java:28)
    at org.codehaus.plexus.archiver.zip.ConcurrentJarCreator$DeferredSupplier.get(ConcurrentJarCreator.java:47)
    at org.apache.commons.compress.archivers.zip.ParallelScatterZipCreator.createDeferred(ParallelScatterZipCreator.java:75)
    at org.apache.commons.compress.archivers.zip.ParallelScatterZipCreator.access$100(ParallelScatterZipCreator.java:54)
    at org.apache.commons.compress.archivers.zip.ParallelScatterZipCreator$1.initialValue(ParallelScatterZipCreator.java:84)
    at org.apache.commons.compress.archivers.zip.ParallelScatterZipCreator$1.initialValue(ParallelScatterZipCreator.java:80)
    at java.lang.ThreadLocal.setInitialValue(ThreadLocal.java:180)
    at java.lang.ThreadLocal.get(ThreadLocal.java:170)
    at org.apache.commons.compress.archivers.zip.ParallelScatterZipCreator$2.call(ParallelScatterZipCreator.java:175)
    ... 4 more
```

I managed to reproduce this locally by comparing the following two commands:

```
1. Runs fine in under 10 seconds
MAVEN_OPTS=-Xmx128m mvn -e org.apache.maven.plugins:maven-source-plugin:2.4:jar
1. Fails already after the third module with source code, after some 7 seconds
MAVEN_OPTS=-Xmx128m mvn -e org.apache.maven.plugins:maven-source-plugin:3.0.0:jar
```

Our project is closed-source, but I verified the issue can also be reproduced on a large Maven build such as the one of the [Jetty](https://github.com/eclipse/jetty.project) project.

---

**Affects:** 3.0.0

**Attachments:**
- [compare-plexus-archiver-revisions.sh](https://issues.apache.org/jira/secure/attachment/12804189/compare-plexus-archiver-revisions.sh) (_2.79 kB_)
- [memory-usage-test.sh](https://issues.apache.org/jira/secure/attachment/12804167/memory-usage-test.sh) (_787 bytes_)
- [memory-usage-test-gc-logs.tgz](https://issues.apache.org/jira/secure/attachment/12804168/memory-usage-test-gc-logs.tgz) (_12.98 kB_)
- [memory-usage-test-gc-logs-maven-3.0.5.tgz](https://issues.apache.org/jira/secure/attachment/12804175/memory-usage-test-gc-logs-maven-3.0.5.tgz) (_12.49 kB_)

**Issue Links:**
- [MSOURCES-98](https://issues.apache.org/jira/browse/MSOURCES-98) Upgrade of 'plexus-archiver' to version 3.3.
 (_**"depends upon"**_)
- [CALCITE-1574](https://issues.apache.org/jira/browse/CALCITE-1574) Memory leak in maven
 (_**"is depended upon by"**_)

3 votes, 6 watchers

**Comment by jira-importer** [comment]
**[Arun Salimath](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=aruns)** commented

I have updated my brew and then mvn however I still use the same maven 3.1.1 version,  (On Mac)

A project that used to be built using maven 3.1.1 with java version 1.6  is not building, error is same as mentioned in this issue.  Note that I was able to build the same project before alright.  Now I am able to build it using   mvn 3.1.1 and Java 1.7 but not with my previous settings.   Also I checked with colleague of mine, on his machine he is able to build it fine  with 3.1.1 and java 1.6 (he did not make any brew update or maven).   Build uses same  maven-source-plugin version  3.0.0  on both mine and colleagues machine. but on mine it fails and on his it succeeds.

**Comment by jira-importer** [comment]
**[Kristian Rosenvold](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=krosenvold)** commented

How much more memory does it actually require, and how many CPU cores do you have ?

**Comment by jira-importer** [comment]
**[Karl Heinz Marbaise](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=khmarbaise)** commented

Can you please check this with Maven 3.0.5 cause i'm investigating memory [usage issue with Maven greater than 3.0.5](https://github.com/khmarbaise/maven-test-project-generator)?

**Comment by jira-importer** [comment]
**[Stephan Schroevers](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=stephan202)** commented

[Karl Heinz Marbaise](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=khmarbaise), that's an interesting finding w.r.t. memory usage of newer Maven versions!

Unfortunately it doesn't seem to be related to this issue, since switching to Maven 3.0.5 doesn't change the test result in any meaningful sense (new GC logs [attached|^memory-usage-test-gc-logs-maven-3.0.5.tgz]):

```
$ mvn -version
Apache Maven 3.0.5 (r01de14724cdef164cd33c7c8c2fe155faf9602da; 2013-02-19 14:51:28+0100)
Maven home: /home/sschroevers/tools/mvn30
Java version: 1.8.0_77, vendor: Oracle Corporation
Java home: /home/sschroevers/tools/jdk1.8.0_77/jre
Default locale: en_US, platform encoding: UTF-8
OS name: "linux", version: "3.19.0-32-generic", arch: "amd64", family: "unix"
$ ./memory-usage-test.sh 
Testing maven-source-plugin:2.4:jar-no-fork with 128m:
Total time: 2.162s
Final Memory: 27M/101M
Testing maven-source-plugin:2.4:jar-no-fork with 256m:
Total time: 2.340s
Final Memory: 30M/214M
Testing maven-source-plugin:2.4:jar-no-fork with 512m:
Total time: 2.181s
Final Memory: 30M/327M
Testing maven-source-plugin:2.4:jar-no-fork with 1024m:
Total time: 2.341s
Final Memory: 32M/429M
Testing maven-source-plugin:2.4:jar-no-fork with 2048m:
Total time: 2.226s
Final Memory: 32M/429M
Testing maven-source-plugin:2.4:jar-no-fork with 4096m:
Total time: 2.234s
Final Memory: 32M/429M
Testing maven-source-plugin:3.0.0:jar-no-fork with 128m:
Total time: 2.608s
Final Memory: 62M/114M
java.lang.OutOfMemoryError: Java heap space
Testing maven-source-plugin:3.0.0:jar-no-fork with 256m:
Total time: 12.324s
Final Memory: 27M/236M
Testing maven-source-plugin:3.0.0:jar-no-fork with 512m:
Total time: 9.124s
Final Memory: 28M/455M
Testing maven-source-plugin:3.0.0:jar-no-fork with 1024m:
Total time: 5.437s
Final Memory: 28M/687M
Testing maven-source-plugin:3.0.0:jar-no-fork with 2048m:
Total time: 5.056s
Final Memory: 29M/884M
Testing maven-source-plugin:3.0.0:jar-no-fork with 4096m:
Total time: 4.386s
Final Memory: 32M/1493M
```

**Comment by jira-importer** [comment]
**[Stephan Schroevers](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=stephan202)** commented

I've traced part of the issue to [plexus-archiver@f6dd097](https://github.com/codehaus-plexus/plexus-archiver/commit/f6dd0976c070440eca0d060cb8baf2b99567b615) (MASSEMBLY-768, replacing `ParallelScatterZipCreator` with `ConcurrentJarCreator`). The attached [^compare-plexus-archiver-revisions.sh] compares that revision with the one just before it, and without that change the OOM does not happen. Script output:

```
Building commons-compress:1.10-SNAPSHOT...
Building plexus-archiver:2.10.1-SNAPSHOT @ f6dd0976c070440eca0d060cb8baf2b99567b615...
Building maven-source-plugin:3.0.1-SNAPSHOT with plexus-archiver:2.10.1-SNAPSHOT...
<snip; the versions-maven-plugin is very verbose>
Checking out jetty.project:jetty-9.3.9.M1 for testing...
Running first test...
Testing org.apache.maven.plugins:maven-source-plugin:3.0.1-SNAPSHOT:jar-no-fork with 128m:
Total time: 3.532 s
Final Memory: 74M/114M
java.lang.OutOfMemoryError: Java heap space
Testing org.apache.maven.plugins:maven-source-plugin:3.0.1-SNAPSHOT:jar-no-fork with 256m:
Total time: 11.871 s
Final Memory: 38M/236M
Testing org.apache.maven.plugins:maven-source-plugin:3.0.1-SNAPSHOT:jar-no-fork with 512m:
Total time: 7.088 s
Final Memory: 38M/453M
Testing org.apache.maven.plugins:maven-source-plugin:3.0.1-SNAPSHOT:jar-no-fork with 1024m:
Total time: 4.175 s
Final Memory: 38M/667M
Testing org.apache.maven.plugins:maven-source-plugin:3.0.1-SNAPSHOT:jar-no-fork with 2048m:
Total time: 3.623 s
Final Memory: 39M/847M
Testing org.apache.maven.plugins:maven-source-plugin:3.0.1-SNAPSHOT:jar-no-fork with 4096m:
Total time: 3.914 s
Final Memory: 42M/1450M
Building plexus-archiver:2.10.1-SNAPSHOT @ bb30b107c78deebf5c3d3227fb8614c03f9275d0...
Running second test...
Testing org.apache.maven.plugins:maven-source-plugin:3.0.1-SNAPSHOT:jar-no-fork with 128m:
Total time: 3.223 s
Final Memory: 38M/112M
Testing org.apache.maven.plugins:maven-source-plugin:3.0.1-SNAPSHOT:jar-no-fork with 256m:
Total time: 2.965 s
Final Memory: 41M/224M
Testing org.apache.maven.plugins:maven-source-plugin:3.0.1-SNAPSHOT:jar-no-fork with 512m:
Total time: 2.685 s
Final Memory: 41M/291M
Testing org.apache.maven.plugins:maven-source-plugin:3.0.1-SNAPSHOT:jar-no-fork with 1024m:
Total time: 2.697 s
Final Memory: 44M/461M
Testing org.apache.maven.plugins:maven-source-plugin:3.0.1-SNAPSHOT:jar-no-fork with 2048m:
Total time: 2.987 s
Final Memory: 40M/708M
Testing org.apache.maven.plugins:maven-source-plugin:3.0.1-SNAPSHOT:jar-no-fork with 4096m:
Total time: 3.174 s
Final Memory: 45M/1220M
```

Note that even before `f6dd097` the run time and memory usage is not on par with 2.4, so there is some additional cause.

I will not look into this further, for now at least, because I need to spend some time on other things.

NB: I also tested with the latest revision of `plexus-archiver`, and that didn't fix the problem. So [plexus-archiver PR #25](https://github.com/codehaus-plexus/plexus-archiver/pull/25) ("_Fix potential resource leak in ConcurrentJarCreator_") doesn't seem related.

**Comment by jira-importer** [comment]
**[Andreas Kohn](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=ankon)** commented

I hit the same issue, and instead of manually patching tried to get maven to use plexus-archiver 3.3:

1. Find all plugins that depend on plexus-archiver:

```
$ mvn dependency:resolve-plugins | grep -E '(Plugin Resolved|Plugin Dependency Resolved: plexus-archiver)' | sed 's,^\[INFO\],,' | perl -0777 -pe 's/\n\s+Plugin Dependency Resolved://igs' | grep plexus-archiver | sort | uniq
 Plugin Resolved: gwt-maven-plugin-2.7.0.jar plexus-archiver-2.6.3.jar
 Plugin Resolved: maven-assembly-plugin-2.6.jar plexus-archiver-3.0.1.jar
 Plugin Resolved: maven-dependency-plugin-2.10.jar plexus-archiver-2.9.jar
 Plugin Resolved: maven-jar-plugin-3.0.0.jar plexus-archiver-3.1.1.jar
 Plugin Resolved: maven-site-plugin-3.3.jar plexus-archiver-1.0.jar
 Plugin Resolved: maven-source-plugin-3.0.0.jar plexus-archiver-3.0.3.jar
 Plugin Resolved: maven-war-plugin-2.6.jar plexus-archiver-2.9.jar
```

2. For of these plugins add/modify the pluginManagement information and add a dependency to org.codehaus.plexus:plexus-archiver:3.3
3. For maven-assembly-plugin and maven-war-plugin: additionally add a dependency to org.codehaus.plexus:plexus-io:2.7.1 to avoid

```
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-war-plugin:2.6:war (default-war) on project activity-engine: Execution default-war of goal org.apache.maven.plugins:maven-war-plugin:2.6:war failed: An API incompatibility was encountered while executing org.apache.maven.plugins:maven-war-plugin:2.6:war: java.lang.NoSuchMethodError: org.codehaus.plexus.components.io.resources.PlexusIoResourceCollection.isConcurrentAccessSupported()Z
```

4. Enjoy working builds again!

So, yes, please do update plexus-archiver :)
